### ISSUE TITLE
Preallocation of matrix memory

### ISSUE THREAD

**Issue by fdkong** [idaholab/moose#7901]
### Description of the enhancement or error report
Preallocation of matrix memory is crucial for good performance for large problems. There are more details at https://www.mcs.anl.gov/petsc/documentation/faq.html

### Identified impact
(i.e. Internal object changes, limited interface changes, public API change, or a list of specific applications impacted)

**Comment by fdkong** [comment]
@permcody, @friedmud, @jwpeterson, @bwspenc,  any comments?

**Comment by bwspenc** [comment]
I agree that we need to straighten things out, and it sounds like we should be able to see some big performance gains when we get this right for contact problems, which I'm excited about (have you figured out what the problem is yet?). We can't turn on the preallocation error messages by default until we can run all of our test suite problems without hitting that error, though, which I imagine will be no small task.

**Comment by jwpeterson** [comment]
@fdkong thanks for getting these numbers.  It is interesting that the over-allocating case runs a bit more quickly, so there is definitely a trade-off present in the two approaches.  As we discussed, I think this should be an (advanced) option which users can select, probably on the `[Problem]` block.

As for your second issue, I believe this is related to @roystgnr's ghosting functor work... it is a way of telling libmesh that certain dofs are coupled when they otherwise would not be, and this information should be used in creating the sparsity pattern.

**Comment by fdkong** [comment]
> I agree that we need to straighten things out, and it sounds like we should be able to see some big performance gains when we get this right for contact problems, which I'm excited about (have you figured out what the problem is yet?). 

I did not figure out yet, and  am still working on.  I think for the second issue, libmesh works fine.  There is something wrong in the function `NonlinearSystem::augmentSparsity` at the MOOSE side.


>We can't turn on the preallocation error messages by default until we can run all of our test suite problems without hitting that error, though, which I imagine will be no small task.

This is a huge task. I am not sure if we are going to fix all problems, @permcody? My first goal is to fix this issue for contact mechanics problems  because I know something about this application and I want to make fieldSplit preconditioner work.

**Comment by friedmud** [comment]
I will think this should be put on hold until someone wants to do it using
the GhostingFunctor stuff.

That is the right way to do it. @roystognr even built in special stuff for
dealing with the fact that not every variable on ghosted dogs is coupled to
every other one (something our stuff has always ignored).

**Comment by friedmud** [comment]
BTW: `need_full_sparsity_pattern` will actually use far MORE RAM!

The SparsityPattern object itself takes a pretty huge amount of memory (and a lot of time) to build.  

Rerun your first stuff on a large problem and use the memory_logger and you'll see the big spike in the beginning when the SparsityPattern object is built.

The over-allocation is preferable to that.  That memory spike could cause your problem to die in memory constrained cases.

**Comment by fdkong** [comment]
> I will think this should be put on hold until someone wants to do it using the GhostingFunctor stuff. 

For two  issues, right? 

>That is the right way to do it. @roystognr even built in special stuff for dealing with the fact that not every variable on ghosted dogs is coupled to every other one (something our stuff has always ignored).

**Comment by fdkong** [comment]
> BTW: need_full_sparsity_pattern will actually use far MORE RAM!
> 
> The SparsityPattern object itself takes a pretty huge amount of memory (and a lot of time) to build.
> 
> Rerun your first stuff on a large problem and use the memory_logger and you'll see the big spike in the beginning when the SparsityPattern object is built.
> 
> The over-allocation is preferable to that. That memory spike could cause your problem to die in memory constrained cases.

I do not understand why this happens?  I implemented  a preallocation algorithm in my in-house code, I did not see any memory issue yet, and also is not slow at all.

**Comment by friedmud** [comment]
@fdkong Yes... the gometric search system will feed the GhostingFunctors the right information.  Then the sparsity pattern should be based on the GhostingFunctors.

**Comment by friedmud** [comment]
@fdkong did you delete a message or something?  I got an email about another message but I don't see it.

It looks like you were questioning about why setting `need_full_sparsity_pattern` will result in increased memory and a slowdown:

1.  The "slowdown" is just that it takes time to build the full sparsity pattern.  That can be quite a LOT of time on very large problems.

2.  The memory increase is due to the creation of the `SparsityPattern` object.  When `need_full_sparsity_pattern = true` a `SparsityPattern` object gets built and completely filled with entries (which takes memory and time).
  With `need_full_sparsity_pattern = false` that's not the case... only the number of nonzeros are tallied up.
  See: https://github.com/libMesh/libmesh/blob/master/src/base/sparsity_pattern.C

**Comment by friedmud** [comment]
In particular: check this: https://github.com/libMesh/libmesh/blob/master/src/base/sparsity_pattern.C#L346

**Comment by fdkong** [comment]
> @fdkong Yes... the gometric search system will feed the GhostingFunctors the right information. Then the sparsity pattern should be based on the GhostingFunctors.

@friedmud, Are GhostingFunctors under development? Or something already mature?

**Comment by fdkong** [comment]
> It looks like you were questioning about why setting need_full_sparsity_pattern will result in increased memory and a slowdown:
> 
>     The "slowdown" is just that it takes time to build the full sparsity pattern. That can be quite a LOT of time on very large problems.
> 
>     The memory increase is due to the creation of the SparsityPattern object. When need_full_sparsity_pattern = true a SparsityPattern object gets built and completely filled with entries (which takes memory and time).
>     With need_full_sparsity_pattern = false that's not the case... only the number of nonzeros are tallied up.
>     See: https://github.com/libMesh/libmesh/blob/master/src/base/sparsity_pattern.C

@friedmud, I understand the concern here.  `SparsityPattern` takes as much memory as a matrix. But I do not need `SparsityPattern` because PETSc does not need it. What PETSc requires is how many nonzeros we have on each row, not the exact locations.

 I possibly like to see the third option. What I need is two correct vectors: `std::vector<dof_id_type> & n_nz` `  std::vector<dof_id_type> & n_oz` without over-allocation and without creating the matrix `SparsityPattern`. The two vectors give me over-allocation without using  `need_full_sparsity_pattern = true`.

**Comment by permcody** [comment]
GhostingFunctors were started at this year's tiger team. They aren't even available in our version of libMesh yet.

**Comment by permcody** [comment]
I plan to meet with @friedmud in Boston at the beginning of November to start hammering out a design for an interface to these new GhostingFunctors. There are several applications for them including this one. We just have to decide what that interface should look like to a MOOSE user.

**Comment by fdkong** [comment]
> I plan to meet with @friedmud in Boston at the beginning of November to start hammering out a design for an interface to these new GhostingFunctors. There are several applications for them including this one. We just have to decide what that interface should look like to a MOOSE user.

@permcody, may I just go ahead to fix prealloaction issue for contact mechanics problems? Or I have to wait for your new interfaces.

I do not think it is necessary to wait. It should be easy migrate things from the old one to this new interface once the new interface is ready to go.

**Comment by permcody** [comment]
If you can fix it without the new interface, great! Don't let this hold you
back.

**Comment by lindsayad** [comment]
@recuero this will be the primary issue I reference for future work on dynamic sparsity patterns

**Comment by gambka** [comment]
I was pointed to this issue based upon behavior I'm seeing with a BISON problem I'm trying to run. The problem has 402120 nonlinear dofs of both `SECOND LAGRANGE` and `FIRST SCALAR` and I'm running on 8 procs. I've ran two versions of the problem, one with thermal and mechanical contact (non-mortar), and one without. In the case with contact the very first `Computing Jacobian` is as follows:

```
Time Step 1, time = 1, dt = 1
 0 Nonlinear |R| = 1.317569e+09

    Computing Jacobian....................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................... [7205.27 s] [  733 MB]
```

However, if I remove thermal and mechanical contact from the problem I get the following:

```
Time Step 1, time = 1, dt = 1
 0 Nonlinear |R| = 1.317569e+09
Currently Executing
    Computing Jacobian                                                                   [ 10.21 s] [  265 MB]
```

All other Jacobian computations for other nonlinear iterations in both problems take ~10 seconds after the initial. That time spent in the initial Computing Jacobian is a significant portion of the computational time in the problem with thermal and mechanical contact.

**Comment by lindsayad** [comment]
I suspect you would run into this issue with mortar as well. However, it would not be in the initial Jacobian evaluation. I had to look twice to verify that you were *not* running with mortar.

This is a Technical Area support issue that we plan to tackle this year. There is an accompanying PETSc issue at https://gitlab.com/petsc/petsc/-/issues/852 that I created which discusses need of contact

**Comment by gambka** [comment]
> I suspect you would run into this issue with mortar as well. However, it would not be in the initial Jacobian evaluation. I had to look twice to verify that you were _not_ running with mortar.

This particular geometric representation I'm using cannot use mortar. Thank you for looking into the issue this year.

**Comment by lindsayad** [comment]
Looking into the skipped test at https://github.inl.gov/ncrc/bison/pull/6072, we have the sparsity pattern wrong even before we apply constraints for the very first Jacobian evaluation. Rough. What's interesting is that I have to run with at least two procs to see the new nonzero error. In serial I do not see it

**Comment by lindsayad** [comment]
This does not directly fix the preallocation issue, but hash table matrix assembly can circumvent it as of https://github.com/idaholab/moose/pull/28979. Hash table matrix assembly issues can be followed in #29069
